apps/shared/iccmaker.c

```diff
@@ -0,0 +1,472 @@
+{
+    if (fabsf(xy[1]) < small) {
+        return AVIF_FALSE;
+    }
+
+    const double factor = 1 / xy[1];
+    XYZ[0] = xy[0] * factor;
+    XYZ[1] = 1;
+    XYZ[2] = (1 - xy[0] - xy[1]) * factor;
+
+    return AVIF_TRUE;
+}
+
+// Computes I = M^-1. Returns false if M seems to be singular.
+static avifBool matInv(const double M[3][3], double I[3][3])
+{
+    double det = M[0][0] * (M[1][1] * M[2][2] - M[2][1] * M[1][2]) - M[0][1] * (M[1][0] * M[2][2] - M[1][2] * M[2][0]) +
+                 M[0][2] * (M[1][0] * M[2][1] - M[1][1] * M[2][0]);
+    if (fabs(det) < small) {
+        return AVIF_FALSE;
+    }
+    det = 1 / det;
+
+    I[0][0] = (M[1][1] * M[2][2] - M[2][1] * M[1][2]) * det;
+    I[0][1] = (M[0][2] * M[2][1] - M[0][1] * M[2][2]) * det;
+    I[0][2] = (M[0][1] * M[1][2] - M[0][2] * M[1][1]) * det;
+    I[1][0] = (M[1][2] * M[2][0] - M[1][0] * M[2][2]) * det;
+    I[1][1] = (M[0][0] * M[2][2] - M[0][2] * M[2][0]) * det;
+    I[1][2] = (M[1][0] * M[0][2] - M[0][0] * M[1][2]) * det;
+    I[2][0] = (M[1][0] * M[2][1] - M[2][0] * M[1][1]) * det;
+    I[2][1] = (M[2][0] * M[0][1] - M[0][0] * M[2][1]) * det;
+    I[2][2] = (M[0][0] * M[1][1] - M[1][0] * M[0][1]) * det;
+
+    return AVIF_TRUE;
+}
+
+// Computes C = A*B
+static void matMul(const double A[3][3], const double B[3][3], double C[3][3])
+{
+    C[0][0] = A[0][0] * B[0][0] + A[0][1] * B[1][0] + A[0][2] * B[2][0];
+    C[0][1] = A[0][0] * B[0][1] + A[0][1] * B[1][1] + A[0][2] * B[2][1];
+    C[0][2] = A[0][0] * B[0][2] + A[0][1] * B[1][2] + A[0][2] * B[2][2];
+    C[1][0] = A[1][0] * B[0][0] + A[1][1] * B[1][0] + A[1][2] * B[2][0];
+    C[1][1] = A[1][0] * B[0][1] + A[1][1] * B[1][1] + A[1][2] * B[2][1];
+    C[1][2] = A[1][0] * B[0][2] + A[1][1] * B[1][2] + A[1][2] * B[2][2];
+    C[2][0] = A[2][0] * B[0][0] + A[2][1] * B[1][0] + A[2][2] * B[2][0];
+    C[2][1] = A[2][0] * B[0][1] + A[2][1] * B[1][1] + A[2][2] * B[2][1];
+    C[2][2] = A[2][0] * B[0][2] + A[2][1] * B[1][2] + A[2][2] * B[2][2];
+}
+
+// Set M to have values of d on the leading diagonal, and zero elsewhere.
+static void matDiag(const double d[3], double M[3][3])
+{
+    M[0][0] = d[0];
+    M[0][1] = 0;
+    M[0][2] = 0;
+    M[1][0] = 0;
+    M[1][1] = d[1];
+    M[1][2] = 0;
+    M[2][0] = 0;
+    M[2][1] = 0;
+    M[2][2] = d[2];
+}
+
+static void swap(double * a, double * b)
+{
+    double tmp = *a;
+    *a = *b;
+    *b = tmp;
+}
+
+// Transpose M
+static void matTrans(double M[3][3])
+{
+    swap(&M[0][1], &M[1][0]);
+    swap(&M[0][2], &M[2][0]);
+    swap(&M[1][2], &M[2][1]);
+}
+
+// Computes y = M.x
+static void vecMul(const double M[3][3], const double x[3], double y[3])
+{
+    y[0] = M[0][0] * x[0] + M[0][1] * x[1] + M[0][2] * x[2];
+    y[1] = M[1][0] * x[0] + M[1][1] * x[1] + M[1][2] * x[2];
+    y[2] = M[2][0] * x[0] + M[2][1] * x[1] + M[2][2] * x[2];
+}
+
+// MD5 algorithm. See https://www.ietf.org/rfc/rfc1321.html#appendix-A.3
+static void computeMD5(uint8_t * data, size_t length)
```

Comment:
Document the input and output of this function.

Document the requirements on `length`. For example, does `length` have to be a multiple of 64?